meta: considerations for new core modules

jasnell opened the pull request:
/cc @AndreasMadsen @evanlucas

Trott commented on COLLABORATOR_GUIDE.md:
```diff
@@ -263,6 +263,20 @@ multiple commits. Commit metadata and the reason for the revert should be
+Semver-minor commits that introduce new core modules should be treated with
+extra care.
+
+* First, the name of the new core module should not conflict with any existing
```
Nit: Remove `First, `.

Trott replied:
There's also singular/plural disagreement in this sentence. Here's what I'd write:

> * The name of the new core module should not conflict with an existing module in the ecosystem unless there is an agreement with the owner of the existing module to transfer ownership.

Trott commented on COLLABORATOR_GUIDE.md:
```diff
@@ -263,6 +263,20 @@ multiple commits. Commit metadata and the reason for the revert should be
+* First, the name of the new core module should not conflict with any existing
+  module in the ecosystem unless an agreement with the owner of those modules
+  is reached to transfer ownership.
+* PRs that introduce new core modules must be landed as semver-minor commits.
```
I think this is unnecessary to state. Any new feature must be landed as semver-minor.  If you're trying to say that even landing as `Experimental` is semver-minor (which I agree it should be), put that in with the statement about it needing to be `Experimental` first.

Trott commented on COLLABORATOR_GUIDE.md:
```diff
@@ -263,6 +263,20 @@ multiple commits. Commit metadata and the reason for the revert should be
+  is reached to transfer ownership.
+* PRs that introduce new core modules must be landed as semver-minor commits.
+* New core modules must be landed initially as `Experimental` and must go
+  through at least one release cycle before being moved out of experimental.
```
The term "release cycle" is probably vague to a lot of people.

Trott replied:
So, yeah, putting my comments on the second and third bullet points together, maybe consolidate them into this single bullet point:

> * New core modules must be landed with a [Stability Index](./doc/api/documentation.md#stability-index) of `Experimental`. (Pull requests proposing new core modules must still be marked `semver-minor`.) New core modules must remain `Experimental` until a `semver-major` release.

Trott commented on COLLABORATOR_GUIDE.md:
```diff
@@ -263,6 +263,20 @@ multiple commits. Commit metadata and the reason for the revert should be
+* New core modules must be landed initially as `Experimental` and must go
+  through at least one release cycle before being moved out of experimental.
+* It is recommended to give PRs introducing new core modules more time for
+  review than the typical 48/72 hour review used for other commits.
```
Suggest a concrete time framer here. I'd go with this:

> * Pull requests introducing new core modules should remain open for one week rather than the usual 48-hour/72-hour review period for other pull requests.

jasnell replied:
I would suggest longer.... maybe even up to a month?

Trott replied:
Maybe phrase it as a minimum? (`open for at least one week`)

BridgeAR replied:
I actually agree with @jasnell as this would probably have already prevented some trouble. Landing a new module should normally not be urgent.
One alternative might also be smth like `At least one week without anyone requesting any further changes.`?
And I would not make it a recommendation but mandatory with a optional fast path through a decision of the CTC/TSC?

addaleax replied:
I think it should be possible for collaborators to pick an appropriate timeframe, even without a fixed rule. One week as a minimum sounds good, but in any case the gist of it will be “if your PR is large/has profound impact, wait until people have had a fair chance to review & assess it”

AndreasMadsen replied:
Please explain why we don't want to go the EPS route. If a new module isn't a major addition I don't know what is.

jasnell replied:
EPS is an ok for speculative large changes but it's really hasn't been that successful of a process over all. Having a concrete pr available, with real code, real tests, real documentation is far easier to review and look at to determine if this is something we want. For instance, Bradley's ESM pr has been a thousand times for valuable than the endless noisy threads in the EPS repo on it. Same goes for the efforts that led to N-API, URL, Inspector, etc. Yes, some discussion in EPS can be helpful but gating whether or not a semver minor can land based on that seems far too unnecessarily heavyweight of a process and, frankly, discourages innovation significantly. I'd definitely be -1 to requiring EPS and have, on several occasions, even considered proposing removing it.

jasnell replied:
@addaleax ... "at least one week" sounds good to me but I would point out that the issue that led to me creating this issue in the first place (Jeremiah's concern that I landed the perf_hooks stuff too fast) meets all of the criteria discussed so far in this thread so I'm not sure if it's sufficient to make everyone comfortable. There needs to be a good balance so that a good compromise can be reached.

cjihrig commented on COLLABORATOR_GUIDE.md:
```diff
@@ -263,6 +263,20 @@ multiple commits. Commit metadata and the reason for the revert should be
+Semver-minor commits that introduce new core modules should be treated with
+extra care.
+
+* First, the name of the new core module should not conflict with any existing
```
I assume there should be an email, GitHub comment, or some other paper trail that shows the agreement with the existing module owner?

jasnell replied:
something along those lines yes.. I'm not sure how prescriptive we need to be with it.

jasnell commented:
Updated based on feedback.

1. @cjihrig ... I clarified that the agreement with the existing module author should be written
2. @Trott and @addaleax ... I've made the time frame at least one week
3. @AndreasMadsen ... I've added a recommendation to use EPS under certain conditions

PTAL

addaleax commented on COLLABORATOR_GUIDE.md:
```diff
@@ -263,6 +263,24 @@ multiple commits. Commit metadata and the reason for the revert should be
+
+It is recommended to give PRs introducing new core modules at least one week
+for review. For new modules that involve significant effort, non-trivial
+additions to Node.js or significant new capabilities, an [EPS][] is
```
teeny tiny nit: EPS is plural, EP is the singular ;)

jasnell replied:
heh... sigh. Should I just change it to "Enhancement Proposal"?

Trott commented on COLLABORATOR_GUIDE.md:
```diff
@@ -263,6 +263,24 @@ multiple commits. Commit metadata and the reason for the revert should be
+Pull requests proposing new core modules must still be marked semver-minor.
+New core modules must remain Experimental until a semver-major release.
+
+It is recommended to give PRs introducing new core modules at least one week
```
Make it more forceful and concise:

> Pull requests introducing new core modules must be left open for at least one week for review.

jasnell replied:
Done.

Trott commented on COLLABORATOR_GUIDE.md:
```diff
@@ -263,6 +263,24 @@ multiple commits. Commit metadata and the reason for the revert should be
+Pull requests introducing new core modules must be left open for at least
+one week for review. For new modules that involve significant effort,
+non-trivial additions to Node.js or significant new capabilities, an
+[Enhancement Proposal][] is recommended but not required.
```
This sentence is confusing to me. (For example, it implies that new modules typically are trivial additions to Node.js.)

How about this?:

> If a new module has a large API surface, requires extensive changes to Node.js core, or adds extraordinary new functionality, open an [Enhancement Proposal][].

mcollina commented on COLLABORATOR_GUIDE.md:
```diff
@@ -263,6 +263,24 @@ multiple commits. Commit metadata and the reason for the revert should be
+
+The name of the new core module should not conflict with any existing 
+module in the ecosystem unless a written agreement with the owner of those
+modules is reached to transfer ownership.
```
We need to add a statement like: "if the new module name is free, a collaborator should park it as soon as possible, linking to the pull request that introduces the new module".

jasnell replied:
oh right, good point

mcollina replied:
I think you still have not added this.

jasnell replied:
oy... heh... one sec

mcollina commented on COLLABORATOR_GUIDE.md:
```diff
@@ -263,6 +263,24 @@ multiple commits. Commit metadata and the reason for the revert should be
+### Introducing New Modules
+
+Semver-minor commits that introduce new core modules should be treated with
+extra care.
```
I would add that they need to land with two CTC (or maybe TSC now) signoff, like any other breaking change.

jasnell replied:
Would that be a should or a must?

mcollina replied:
I think that's a must. It's serious business: we are adding new API that we will need to maintain to core. Removing those is hard, so the CTC should be notified to weight in.

AndreasMadsen commented:
> EPS is an ok for speculative large changes but it's really hasn't been that successful of a process over all. Having a concrete pr available, with real code, real tests, real documentation is far easier to review and look at to determine if this is something we want. For instance, Bradley's ESM pr has been a thousand times for valuable than the endless noisy threads in the EPS repo on it. Same goes for the efforts that led to N-API, URL, Inspector, etc. Yes, some discussion in EPS can be helpful but gating whether or not a semver minor can land based on that seems far too unnecessarily heavyweight of a process and, frankly, discourages innovation significantly. I'd definitely be -1 to requiring EPS and have, on several occasions, even considered proposing removing it.

I guessed that was the reason. Personally, I disagree, when I look at PEP (Python Enhancement Proposal) I'm always impressed by the level of detail and thought that goes into pretty much any major feature. It is super helpful to learn about the thought process that went into the feature. I think we can learn a lot from them.

But if EPS is something we want to discourage, then we should change the EPS guide lines or deprecate the approach completely. But a minimum requirement for any new major feature should be to put it on the CTC agenda, even if there are no disagreements. That would have made any concerns in this cause quite obvious. This is effectively equivalent to the EPS approach, but with less documentation and more code.

jasnell commented:
@AndreasMadsen ... that's fair :-)

We've been actively moving away from putting things on the `ctc-agenda` label in favor of a more async decision making process using `ctc-review` and only putting exceptions on `ctc-agenda`. It's been working rather well so far. So how about this, we keep the minimum one week, we require two CTC members to sign off (as @mcollina suggests) and we require that the issue is flagged with `ctc-review` during that time....

(Of course, that's in addition to it landing as Experimental)

jasnell commented:
Ok, take a look now! I think I've captured your feedback @AndreasMadsen and @mcollina :-)

jasnell commented:
Landed in 397dbab

targos commented:
Should we notify all collaborators about this change?

jasnell commented:
ping @nodejs/collaborators ... fyi

jkrems commented on COLLABORATOR_GUIDE.md:
```diff
@@ -263,6 +263,32 @@ multiple commits. Commit metadata and the reason for the revert should be
+module in the ecosystem unless a written agreement with the owner of those
+modules is reached to transfer ownership.
+
+If the new module name is free, a Collaborator should register a placeholder
```
Should we commit to adding (or allowing for) polyfills to be published for older versions of node under that name? Or is that out of scope for this generic document and would be discussed case-by-case?

jasnell replied:
I'd say that's out of scope
